#Project 1 

#Part 1

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


##Section 2.3

In [2]:
import pandas as pd

In [3]:
hd = pd.read_csv('/content/drive/My Drive/HDS/CMPS 260/Data/happiness_2017.csv')

In [4]:
hd.head()

Country  ... Confidence in national government
0       Norway  ...                          0.717160
1      Denmark  ...                          0.572353
2      Iceland  ...                          0.365042
3  Switzerland  ...                          0.819707
4      Finland  ...                          0.597539

[5 rows x 14 columns]

###Q-1: How many unique regions are there?

In [5]:
hd1 = hd.groupby('Region').mean()
hd1

Rank  ...  Confidence in national government
Region                                          ...                                   
Central and Eastern Europe           64.176471  ...                           0.342364
Commonwealth of Independent States   76.583333  ...                           0.524538
East Asia                            60.000000  ...                           0.382799
Latin America and Caribbean          44.894737  ...                           0.372781
Middle East and North Africa         68.588235  ...                           0.361473
North America and ANZ                 9.333333  ...                           0.482986
South Asia                          103.000000  ...                           0.601951
Southeast Asia                       72.714286  ...                           0.795365
Sub-Saharan Africa                  116.606061  ...                           0.593257
Western Europe                       23.900000  ...                           0.508622

[10 rows x 12 columns]

In [6]:
hd1.shape

(10, 12)

###Q-2: There are ______ countries in Western Europe and ______countries in Sub-Saharan Africa.

In [7]:
hd2 = hd.groupby('Region').count()['Country']
hd2

Region
Central and Eastern Europe            17
Commonwealth of Independent States    12
East Asia                              6
Latin America and Caribbean           19
Middle East and North Africa          17
North America and ANZ                  3
South Asia                             6
Southeast Asia                         7
Sub-Saharan Africa                    33
Western Europe                        20
Name: Country, dtype: int64

In [8]:
hd2.loc['Western Europe']

20

In [9]:
hd2.loc['Sub-Saharan Africa']

33

###Q-3: The sum of all happiness in South Asia is____ and the sum of all happiness in Western Europe is_____

In [10]:
hd3 = hd.groupby('Region').sum()['HappinessScore'].round(3)
hd3

Region
Central and Eastern Europe             93.869
Commonwealth of Independent States     63.019
East Asia                              33.880
Latin America and Caribbean           114.373
Middle East and North Africa           92.187
North America and ANZ                  21.591
South Asia                             27.388
Southeast Asia                         37.475
Sub-Saharan Africa                    137.097
Western Europe                        134.968
Name: HappinessScore, dtype: float64

In [11]:
hd3.loc['South Asia']

27.388

In [12]:
hd3.loc['Western Europe']

134.968

###Q-4: The average happiness score for Latin America and Caribbean is ____



In [13]:
hd4 = hd.groupby('Region').mean()['HappinessScore'].round(3)
hd4

Region
Central and Eastern Europe            5.522
Commonwealth of Independent States    5.252
East Asia                             5.647
Latin America and Caribbean           6.020
Middle East and North Africa          5.423
North America and ANZ                 7.197
South Asia                            4.565
Southeast Asia                        5.354
Sub-Saharan Africa                    4.154
Western Europe                        6.748
Name: HappinessScore, dtype: float64

In [14]:
hd4.loc['Latin America and Caribbean']

6.02

###Q-5: What is the happiest country in East Asia?

In [15]:
hd5 = hd.loc[hd.Region == "East Asia"][['Region','Country', 'HappinessScore']].sort_values(by = 'HappinessScore', ascending = False)
hd5

Region                   Country  HappinessScore
31  East Asia  Taiwan Province of China           6.422
46  East Asia                     Japan           5.920
51  East Asia               South Korea           5.838
64  East Asia   Hong Kong S.A.R., China           5.472
72  East Asia                     China           5.273
90  East Asia                  Mongolia           4.955

###Q-6: What does your spreadsheet show for the population of the United States? What does the countries of the world sheet show for the United States?

In [16]:
wd = pd.read_csv('/content/drive/My Drive/HDS/CMPS 260/Data/world_countries.csv')
wd.head()

Country Code                Region  ...  Agriculture  Industry  Service
0     Afghanistan  AFG  ASIA (EX. NEAR EAST)  ...        0.380     0.240    0.380
1         Albania  ALB        EASTERN EUROPE  ...        0.232     0.188    0.579
2         Algeria  DZA       NORTHERN AFRICA  ...        0.101     0.600    0.298
3  American Samoa  ASM               OCEANIA  ...          NaN       NaN      NaN
4         Andorra  AND        WESTERN EUROPE  ...          NaN       NaN      NaN

[5 rows x 21 columns]

In [17]:
new_hd = pd.merge(hd, wd, on='Country', how = 'outer')
new_hd.head()

Country        Region_x  Rank  ...  Agriculture  Industry  Service
0       Norway  Western Europe   1.0  ...        0.021     0.415    0.564
1      Denmark  Western Europe   2.0  ...        0.018     0.246    0.735
2      Iceland  Western Europe   3.0  ...        0.086     0.150    0.765
3  Switzerland  Western Europe   4.0  ...        0.015     0.340    0.645
4      Finland  Western Europe   5.0  ...        0.028     0.295    0.676

[5 rows x 34 columns]

In [18]:
new_hd.loc[new_hd.Country == "United States"][['Country', 'Population']]

Country  Population
12  United States         NaN

In [19]:
wd.loc[wd.Country == 'United States of America'][['Country', 'Population']]

Country  Population
212  United States of America   298444215

####In the `merge()` function we used in question 6, I included the argument `how = outer` which returns a join over the union of the input column Country. This argument fills in all missing values with NANs. So countries present in happiness_2017 but not present in world_countries have NAN in their Population column. Here, in happiness_2017 America is listed as 'United States' whereas in world_countries it is listed as 'United States of America'.Hence, the two are not considered the same thing. Therefore, when we search up the population for 'United States' in the merged dataframe, we get NAN. 

###Q-7: Which of the following countries are NOT in the world countries spreadsheet?

In [20]:
wd.loc[wd.Code == 'KOS']

Empty DataFrame
Columns: [Country, Code, Region, Population, Area, Pop. Density, Coastline, Net migration, Infant mortality, GDP, Literacy, Phones, Arable, Crops, Other, Climate, Birthrate, Deathrate, Agriculture, Industry, Service]
Index: []

In [21]:
wd.loc[wd.Code == 'PSE']

Empty DataFrame
Columns: [Country, Code, Region, Population, Area, Pop. Density, Coastline, Net migration, Infant mortality, GDP, Literacy, Phones, Arable, Crops, Other, Climate, Birthrate, Deathrate, Agriculture, Industry, Service]
Index: []

In [22]:
wd.loc[wd.Code == 'PLW']

Country Code   Region  ...  Agriculture  Industry  Service
155   Palau  PLW  OCEANIA  ...        0.062      0.12    0.818

[1 rows x 21 columns]

In [23]:
wd.loc[wd.Code == 'CIV']

Country Code              Region  ...  Agriculture  Industry  Service
49  Côte d'Ivoire  CIV  SUB-SAHARAN AFRICA  ...        0.279     0.171     0.55

[1 rows x 21 columns]

No observation is found for Kosovo and Palestine

###Q-8: The average population is __ for East Asia.

In [24]:
new_hd1 = new_hd.loc[new_hd.Region_x == 'East Asia'][['Region_x', 'Country','Population']]
new_hd1

Region_x                   Country    Population
31  East Asia  Taiwan Province of China           NaN
46  East Asia                     Japan  1.274636e+08
51  East Asia               South Korea           NaN
64  East Asia   Hong Kong S.A.R., China           NaN
72  East Asia                     China  1.313974e+09
90  East Asia                  Mongolia  2.832224e+06

We can see that we are missing population values for Taiwan Province of China, South Kore and Hong Kong S.A.R., China. Lets first remove white space at both ends in the column Country using `strip()`

In [25]:
wd.Country = wd.Country.str.strip()

Now, we search up the index for the country in happiness dataset. Then we find what the Country column in world countries dataset has as the name of that country. Then we change the name in happiness dataset to the name present in world countries dataset.

In [26]:
hd.loc[hd.Country == "Taiwan Province of China"]

Country  ... Confidence in national government
31  Taiwan Province of China  ...                          0.322822

[1 rows x 14 columns]

In [27]:
wd.loc[wd.Code == "TWN"]

Country Code                Region  ...  Agriculture  Industry  Service
196  Taiwan  TWN  ASIA (EX. NEAR EAST)  ...        0.018     0.259    0.723

[1 rows x 21 columns]

In [28]:
hd_changed = hd.copy()
hd_changed.loc[31, 'Country'] = "Taiwan"
hd_changed[hd_changed.Country == 'Taiwan']

Country     Region  ...  Negative affect  Confidence in national government
31  Taiwan  East Asia  ...         0.114123                           0.322822

[1 rows x 14 columns]

In [29]:
hd.loc[hd.Country == "South Korea"]

Country     Region  ...  Negative affect  Confidence in national government
51  South Korea  East Asia  ...         0.234826                           0.358582

[1 rows x 14 columns]

In [30]:
wd.loc[wd.Code == "KOR"]

Country Code                Region  ...  Agriculture  Industry  Service
109  Korea, South  KOR  ASIA (EX. NEAR EAST)  ...        0.033     0.403    0.563

[1 rows x 21 columns]

In [31]:
hd_changed.loc[51, 'Country'] = "Korea, South"
hd_changed[hd_changed.Country == 'Korea, South']

Country     Region  ...  Negative affect  Confidence in national government
51  Korea, South  East Asia  ...         0.234826                           0.358582

[1 rows x 14 columns]

In [32]:
hd.loc[hd.Country == "Hong Kong S.A.R., China"]

Country  ... Confidence in national government
64  Hong Kong S.A.R., China  ...                          0.545041

[1 rows x 14 columns]

In [33]:
wd.loc[wd.Code == "HKG"]

Country Code                Region  ...  Agriculture  Industry  Service
90  Hong Kong  HKG  ASIA (EX. NEAR EAST)  ...        0.001     0.092    0.906

[1 rows x 21 columns]

In [34]:
hd_changed.loc[64, 'Country'] = "Hong Kong"
hd_changed[hd_changed.Country == 'Hong Kong']

Country     Region  ...  Negative affect  Confidence in national government
64  Hong Kong  East Asia  ...         0.200593                           0.545041

[1 rows x 14 columns]

In [35]:
new_merged = pd.merge(hd_changed, wd, on='Country', how = 'outer')
new_merged1 = new_merged.loc[new_hd.Region_x == 'East Asia'][['Region_x', 'Country','Population']]
new_merged1

Region_x       Country    Population
31  East Asia        Taiwan  2.303609e+07
46  East Asia         Japan  1.274636e+08
51  East Asia  Korea, South  4.884682e+07
64  East Asia     Hong Kong  6.940432e+06
72  East Asia         China  1.313974e+09
90  East Asia      Mongolia  2.832224e+06

In [36]:
new_merged1.groupby('Region_x').mean().astype(str)

Population
Region_x              
East Asia  253848815.0

###Q-9: The largest population in Latin America and Caribbean

In [37]:
new_hd2 = new_hd.loc[new_hd.Region_x == "Latin America and Caribbean"][['Region_x', 'Country', 'Population']].sort_values(by = 'Population', ascending = False).dropna()
new_hd2

Region_x             Country   Population
20   Latin America and Caribbean              Brazil  188078227.0
23   Latin America and Caribbean              Mexico  107449525.0
33   Latin America and Caribbean            Colombia   43593035.0
22   Latin America and Caribbean           Argentina   39921833.0
57   Latin America and Caribbean                Peru   28302603.0
18   Latin America and Caribbean               Chile   16134219.0
40   Latin America and Caribbean             Ecuador   13547510.0
26   Latin America and Caribbean           Guatemala   12293545.0
78   Latin America and Caribbean  Dominican Republic    9183984.0
53   Latin America and Caribbean             Bolivia    8989046.0
132  Latin America and Caribbean               Haiti    8308504.0
83   Latin America and Caribbean            Honduras    7326496.0
41   Latin America and Caribbean         El Salvador    6822378.0
39   Latin America and Caribbean           Nicaragua    5570129.0
10   Latin America and Caribbean          Costa Rica    4075261.0
27   Latin America and Caribbean             Uruguay    3431932.0
28   Latin America and Caribbean              Panama    3191319.0
69   Latin America and Caribbean             Jamaica    2758124.0

###Q-10: The smallest population in the Middle East and North Africa region.

In [38]:
new_hd3 = new_hd.loc[new_hd.Region_x == "Middle East and North Africa"][['Region_x', 'Country', 'Population']].sort_values(by = 'Population').dropna()
new_hd3

Region_x               Country  Population
38   Middle East and North Africa               Bahrain    698585.0
36   Middle East and North Africa                Kuwait   2418393.0
19   Middle East and North Africa  United Arab Emirates   2602713.0
80   Middle East and North Africa               Lebanon   3874050.0
62   Middle East and North Africa                 Libya   5900754.0
67   Middle East and North Africa                Jordan   5906760.0
9    Middle East and North Africa                Israel   6352117.0
92   Middle East and North Africa               Tunisia  10175014.0
133  Middle East and North Africa                 Yemen  21456188.0
107  Middle East and North Africa                  Iraq  26783383.0
34   Middle East and North Africa          Saudi Arabia  27019731.0
48   Middle East and North Africa               Algeria  32930091.0
76   Middle East and North Africa               Morocco  33241259.0
98   Middle East and North Africa                  Iran  68688433.0
63   Middle East and North Africa                Turkey  70413958.0
94   Middle East and North Africa                 Egypt  78887007.0

###Q-11: Write down two questions of your own, that you can explore with the combined data set.

####a. Does literacy play a role in perceptions of corruptions?

In [39]:
CL = new_hd[['Country','Perceptions of corruption', 'Literacy']].dropna()
CL.head()


Country  Perceptions of corruption  Literacy
0       Norway                   0.249711     100.0
1      Denmark                   0.181148     100.0
2      Iceland                   0.726845      99.9
3  Switzerland                   0.316183      99.0
4      Finland                   0.192413     100.0

In [40]:
CL['Literacy'].describe()

count    112.000000
mean      81.419643
std       21.366750
min       17.600000
25%       69.775000
50%       91.550000
75%       98.525000
max      100.000000
Name: Literacy, dtype: float64

Using the 1st quartile and 3rd quartile values as boundaries I will create a new column called 'Literacy_Category' which will categorize each country into either 'Low Literacy, 'Moderate Literacy' or 'High Literacy' based on the literacy boundaries.

In [41]:
def literacy_category(literacy):
  if literacy <= 69.775:
    return 'Low Literacy'
  elif literacy >= 98.525:
    return 'High Literacy'
  else:
    return 'Moderate Literacy'

In [42]:
CL['Literacy Category'] = CL['Literacy'].apply(literacy_category)
CL.head()

Country  Perceptions of corruption  Literacy Literacy Category
0       Norway                   0.249711     100.0     High Literacy
1      Denmark                   0.181148     100.0     High Literacy
2      Iceland                   0.726845      99.9     High Literacy
3  Switzerland                   0.316183      99.0     High Literacy
4      Finland                   0.192413     100.0     High Literacy

In [43]:
CL1 = CL[['Literacy Category', 'Perceptions of corruption']]
CL1 = CL1.groupby('Literacy Category').mean().sort_values(by='Perceptions of corruption')
CL1

Perceptions of corruption
Literacy Category                           
High Literacy                       0.598176
Moderate Literacy                   0.771060
Low Literacy                        0.783124

####Answer: Here we can see that literacy is indirectly proportional to perceptions of corruption. Countries with high literacy have the lowest perception of corruption whereas countries with low literacy have the highest perception of corruption, which suggests that literacy does play a part in people's perception of presence of corruption.  

####b.Which region has the highest average GDP. Does this region also have the highest average happiness score?

We are using median as a middle measure for GDP so that extreme values do not majorly affect it.

In [44]:
GH = new_hd[['Region_x', 'GDP','HappinessScore']]
GH = GH.groupby('Region_x').median().sort_values(by = 'GDP', ascending = False)
GH

GDP  HappinessScore
Region_x                                                   
Western Europe                      27600.0          6.9210
North America and ANZ               25300.0          7.2840
Central and Eastern Europe          10850.0          5.3950
Middle East and North Africa         6550.0          5.3360
Latin America and Caribbean          5550.0          6.1680
East Asia                            5000.0          5.6550
Commonwealth of Independent States   3450.0          5.4015
Southeast Asia                       3200.0          5.2620
South Asia                           2000.0          4.5240
Sub-Saharan Africa                   1200.0          4.1800

####Answer: Western Europe has the largest GDP. Now lets see if it has the highest happiness score too. 

In [45]:
GH1 = GH.sort_values(by = 'HappinessScore', ascending = False)
GH1

GDP  HappinessScore
Region_x                                                   
North America and ANZ               25300.0          7.2840
Western Europe                      27600.0          6.9210
Latin America and Caribbean          5550.0          6.1680
East Asia                            5000.0          5.6550
Commonwealth of Independent States   3450.0          5.4015
Central and Eastern Europe          10850.0          5.3950
Middle East and North Africa         6550.0          5.3360
Southeast Asia                       3200.0          5.2620
South Asia                           2000.0          4.5240
Sub-Saharan Africa                   1200.0          4.1800

####Here we see that Western Europe doesnt have the highest Happiness score despite having the highest GDP. However, it comes up second in Happiness after North America and ANZ. 

##Section 2.4

In [46]:
h = pd.read_csv('/content/drive/My Drive/HDS/CMPS 260/Data/happiness_2012.csv')
h.head()

country  ...  gini of household income reported in Gallup, by wp5-year
0      Denmark  ...                                           0.365274       
1       Norway  ...                                           0.318804       
2  Switzerland  ...                                           0.308629       
3  Netherlands  ...                                           0.352454       
4       Sweden  ...                                           0.319790       

[5 rows x 20 columns]

In [47]:
h.rename(columns={'country': 'Country'}, inplace=True)
new_h = pd.merge(h, hd, on='Country', how = 'outer')[['Country','Rank_x', 'Happiness Score', 'Rank_y', 'HappinessScore']]
new_h.rename(columns={'Rank_x': 'Rank 2012', 'Happiness Score':'HappinessScore 2012', 'Rank_y':'Rank 2017','HappinessScore':'HappinessScore 2017'}, inplace=True)
new_h = new_h.set_index('Country')
new_h.head()

Rank 2012  HappinessScore 2012  Rank 2017  HappinessScore 2017
Country                                                                    
Denmark            1.0                7.693        2.0                7.522
Norway             2.0                7.655        1.0                7.537
Switzerland        3.0                7.650        4.0                7.494
Netherlands        4.0                7.512        6.0                7.377
Sweden             5.0                7.480        9.0                7.284

Making a new column that calculates change in rank from 2017 to 2012. 

In [48]:
new_h['Change in Rank (2017-2012)'] = new_h['Rank 2017'] - new_h['Rank 2012']
new_h.head()

Rank 2012  ...  Change in Rank (2017-2012)
Country                 ...                            
Denmark            1.0  ...                         1.0
Norway             2.0  ...                        -1.0
Switzerland        3.0  ...                         1.0
Netherlands        4.0  ...                         2.0
Sweden             5.0  ...                         4.0

[5 rows x 5 columns]

###Q-1: What is the name of the country with the largest positive change in their happiness rank? What about the largest negative change?

In [49]:
new_h.sort_values(by ='Change in Rank (2017-2012)', ascending = False).head()

Rank 2012  ...  Change in Rank (2017-2012)
Country              ...                            
Albania        59.0  ...                        41.0
Ghana          81.0  ...                        40.0
Ukraine        82.0  ...                        40.0
Zimbabwe       95.0  ...                        33.0
Cyprus         31.0  ...                        29.0

[5 rows x 5 columns]

In [50]:
new_h.sort_values(by ='Change in Rank (2017-2012)').head()

Rank 2012  ...  Change in Rank (2017-2012)
Country              ...                            
Bulgaria      133.0  ...                       -37.0
Bahrain        74.0  ...                       -35.0
Nepal         124.0  ...                       -34.0
Latvia         83.0  ...                       -33.0
Hungary       102.0  ...                       -33.0

[5 rows x 5 columns]

###Q-3: What is the name of the country with the largest positive change in their happiness score? What about the largest negative change? 

In [51]:
new_h['Change in HappinessScore (2017-2012)'] = new_h['HappinessScore 2017'] - new_h['HappinessScore 2012']
new_h.head()

Rank 2012  ...  Change in HappinessScore (2017-2012)
Country                 ...                                      
Denmark            1.0  ...                                -0.171
Norway             2.0  ...                                -0.118
Switzerland        3.0  ...                                -0.156
Netherlands        4.0  ...                                -0.135
Sweden             5.0  ...                                -0.196

[5 rows x 6 columns]

In [52]:
new_h.sort_values('Change in HappinessScore (2017-2012)', ascending = False).head()

Rank 2012  ...  Change in HappinessScore (2017-2012)
Country              ...                                      
Nepal         124.0  ...                                 0.806
Latvia         83.0  ...                                 0.804
Romania        85.0  ...                                 0.792
Bahrain        74.0  ...                                 0.775
Bulgaria      133.0  ...                                 0.733

[5 rows x 6 columns]

In [53]:
new_h.sort_values('Change in HappinessScore (2017-2012)').head()

Rank 2012  ...  Change in HappinessScore (2017-2012)
Country              ...                                      
Ghana          81.0  ...                                -0.971
Ukraine        82.0  ...                                -0.961
Zimbabwe       95.0  ...                                -0.952
Albania        59.0  ...                                -0.906
Cyprus         31.0  ...                                -0.845

[5 rows x 6 columns]

###Use the final data frame you constructed above to determine the number of countries included in the 2012 dataset but not in the 2017 dataset, as well as the number of countries included in the 2017 dataset but not in the 2012 dataset

In [54]:
merged_h = pd.merge(h, hd, on='Country', how = 'inner')
merged_h.shape

(126, 33)

This means that 126 countries were in both data sets.

In [55]:
new_h.shape

(156, 6)

This means that the union of the two data sets have 156 countries. 

In [56]:
h.shape 

(142, 20)

This means that 142 countries were in the happiness 2012.

In [57]:
hd.shape

(140, 14)

This means that 140 countries were in happiness 2017. Now, subtracting the number of countries present in both data sets from total number of countries in happiness 2012. 

In [58]:
142-126

16

Therefore, 16 countries were present in happiness 2012 but not in happiness 2017. Now, subtracting the number of countries present in both data sets from total number of countries in happiness 2017.

In [59]:
140-126

14

Therefore, 14 countries were present in happiness 2017 but not in happiness 2012. Now, we are checking if the intersection, countries in happiness 2012, and countries in happiness 2017 add up to the union value.

In [60]:
126+16+14

156

Yes! There were 156 countries in the union table.

#Part 2:


In [ ]:
from google.colab import files
uploaded = files.upload()

In [ ]:
import io
df = pd. read_csv(io.BytesIO(uploaded['Oil_and_Gas_Summary_Production_Data__1967-1999.csv']))

In [ ]:
df.head()

####Oil and gas Production Data from 1967 to 1999 seems pretty interesting. It was from the 2000s when the awareness about environmental sustainability outraged. Before that although environmental consciousness was in the frame, but the production of fossil fuel was the top priority. Alternatives to Fossil Fuel were not quite popular. We wanted to analyze the trend of production of Oil and Gas right before the era of environmental sustainability awareness and consciousness. Active well, Inactive wells and Water production are some of the familiar topics we had heard about oil and gas production but Injection well, Disposal well and Self-use well are some of the new topics we found through this data set. We want to research on the correlation between all these aspects of the data set and their overall impact on the production.



####Qn1: Which production year had the highest number of active oil wells?

In [ ]:
df1= df[['Production Year', 'Active Oil Wells']]
df1.groupby('Production Year').mean().sort_values(by = 'Active Oil Wells', ascending = False).head()

####Answer: The highest number of active oil wells were present in 1968. 

####Qn2: Which county had the highest amount of water production during 1968?

In [ ]:
df2 = df[['Production Year', 'County','Water produced, bbl']]
df2.loc[df2['Production Year'] == 1968].sort_values(by = 'Water produced, bbl', ascending = False).head()

####Cattaraugus had the highest amount of water production during 1968.

In [ ]:
df.shape

In [ ]:
df.dropna().shape


Since, we have a lot more than 1000 observations even after dropping all rows with NAN values, we can easily use this dataset for further analysis. Also, there are no ethical concerns related to the information provided in this dataset. 